File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyNodeOnlyValue(node *Node) *Node {
    if node == nil {
        return nil;
    }
    
    return &Node { Val: node.Val };
}

func copyRandomList(head *Node) *Node {
    mapToClone := make(map[*Node]*Node);
    clonedHead := copyNodeOnlyValue(head);
    mapToClone[head] = clonedHead;
    curCloned := clonedHead;
    for head != nil {
        if mapToClone[head.Next] == nil {
            mapToClone[head.Next] = copyNodeOnlyValue(head.Next);
        }
        curCloned.Next = mapToClone[head.Next];
        if mapToClone[head.Random] == nil {
            mapToClone[head.Random] = copyNodeOnlyValue(head.Random);
        }
        curCloned.Random = mapToClone[head.Random];
        curCloned = mapToClone[head.Next];
        head = head.Next;
    }
    return clonedHead;
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

import java.util.HashMap;

class Solution {
    public Node copyNodeOnlyValue(Node node) {
        if (node == null) {
            return null;
        }
        return new Node(node.val);
    }

    public Node copyRandomList(Node head) {
        HashMap<Node, Node> mapToClone = new HashMap<>();
        Node clonedHead = copyNodeOnlyValue(head);
        mapToClone.put(head, clonedHead);
        Node curClone = clonedHead;
        while (head != null) {
            if (!mapToClone.containsKey(head.next)) {
                mapToClone.put(head.next, copyNodeOnlyValue(head.next));
            }
            curClone.next = mapToClone.get(head.next);
            if (!mapToClone.containsKey(head.random)) {
                mapToClone.put(head.random, copyNodeOnlyValue(head.random));
            }
            curClone.random = mapToClone.get(head.random);
            curClone = curClone.next;
            head = head.next;
        }
        return clonedHead;
    }
}